srtcore/sync.cpp

```diff
@@ -0,0 +1,168 @@
+﻿#include <iomanip>
+#include <math.h>
+#include <stdexcept>
+#include "sync.h"
+#include "udt.h"
+#include "srt_compat.h"
+
+#if defined(_WIN32)
+#define TIMING_USE_QPC
+#include "win/wintime.h"
+#include <sys/timeb.h>
+#elif defined(OSX) || (TARGET_OS_IOS == 1) || (TARGET_OS_TV == 1)
+#define TIMING_USE_MACH_ABS_TIME
+#include <mach/mach_time.h>
+#elif defined(_POSIX_MONOTONIC_CLOCK) && _POSIX_TIMERS > 0
+#define TIMING_USE_MONOTONIC_CLOCK
+#endif
+
+namespace srt
+{
+namespace sync
+{
+
+void rdtsc(uint64_t& x)
+{
+#ifdef IA32
+    uint32_t lval, hval;
+    // asm volatile ("push %eax; push %ebx; push %ecx; push %edx");
+    // asm volatile ("xor %eax, %eax; cpuid");
+    asm volatile("rdtsc" : "=a"(lval), "=d"(hval));
+    // asm volatile ("pop %edx; pop %ecx; pop %ebx; pop %eax");
+    x = hval;
+    x = (x << 32) | lval;
+#elif defined(IA64)
+    asm("mov %0=ar.itc" : "=r"(x)::"memory");
+#elif defined(AMD64)
+    uint32_t lval, hval;
+    asm("rdtsc" : "=a"(lval), "=d"(hval));
+    x = hval;
+    x = (x << 32) | lval;
+#elif defined(TIMING_USE_QPC)
+    // This function should not fail, because we checked the QPC
+    // when calling to QueryPerformanceFrequency. If it failed,
+    // the m_bUseMicroSecond was set to true.
+    QueryPerformanceCounter((LARGE_INTEGER*)&x);
+#elif defined(TIMING_USE_MACH_ABS_TIME)
+    x = mach_absolute_time();
+#else
+    // use system call to read time clock for other archs
+    timeval t;
+    gettimeofday(&t, 0);
```

Comment:
Should be using clock_gettime(CLOCK_MONOTONIC,..) if ENABLE_MONOTONIC_CLOCK is defined.

Reply:
Added monotonic clock.